lib/rules/no-unused-vars.js

```diff
@@ -1153,17 +1153,22 @@ module.exports = {
             }
 
             // skip fix when variable is reassigned
-            if (writeReferences.length > 1 && parentType !== "AssignmentPattern") {
+            if (allWriteReferences.length > 1 && parentType !== "AssignmentPattern") {
+                return null;
+            }
+
+            // skip fix when variable in assignment pattern is reassigned
+            if (allWriteReferences.length > 2 && parentType === "AssignmentPattern") {
                 return null;
             }
```

Comment:
When assignment pattern represents default param value, it has only one write ref for that assignment. For example, in this case there should be no suggestion to remove `x`:

```js
function foo(x = 1) {
    x = 2;
}
```

Reply:
There's also a bug when params with default values are being removed:

```js
function foo(x = 1) {} // fixed to: function foo( = 1) {}
```

Reply:
I think we can simplify this by just checking if all references have the same `range[0]` as `id`.

Reply:
>I think we can simplify this by just checking if all references have the same range[0] as id.

This way aren't we leaving suggestion for following code: 
```js
function foo(x = 1) {}  // no suggestion for x
```
is this the intention?

Though i tried a condition to skip fix for `function foo(x = 1) {x = 2;}` please check if this seems correct.

Reply:
> This way aren't we leaving suggestion for following code:
> 
> ```js
> function foo(x = 1) {}  // no suggestion for x
> ```

Yes, this code should have a suggestion to remove `x = 1`.

Reply:
I was thinking about something like this: https://github.com/eslint/eslint/pull/18352#discussion_r1864932728